### PR TITLE
5 pattern for java workload monitoring

### PR DESCRIPTION
**Pull request by freschri**
**Note: this PR depends on
https://github.com/aws-observability/aws-observability-accelerator/pull/15 and on
https://github.com/aws-quickstart/cdk-eks-blueprints/pull/784**
some changes in here are also in https://github.com/aws-observability/cdk-aws-observability-accelerator/pull/65

### CODE REVIEW

File: cdk.json [hunk 1]
```diff
@@ -17,12 +17,23 @@
...
+    "grafanaGitOpsConfig": {
+      "bootstrapRepo": {
+        "repoUrl": "https://github.com/aws-observability/aws-observability-accelerator",
+        "name": "grafana-dashboards",
+        "targetRevision": "main",
+        "path": "./artifacts/grafana-operator-manifests/eks/infrastructure"
+      },
+      "fluxTargetNamespace": "grafana-operator",
+      "bootstrapValues": {
+        "GRAFANA_CLUSTER_DASH_URL" : "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/infrastructure/cluster.json",
```
Comment (shapirov103) [inline cdk.json:29]: Is there a reason why this is an FQDN URL which happens to point to the same repo as the bootstrap repo? Is there a way to use the dashboard location based off the relative path in the repo and branch that is referenced here?

File: cdk.json [hunk 2]
```diff
@@ -17,12 +17,23 @@
     ]
   },
   "context": {
-    "cluster.dashboard.url": "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/infrastructure/cluster.json",
-    "kubelet.dashboard.url": "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/infrastructure/kubelet.json",
-    "namespaceworkloads.dashboard.url": "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/infrastructure/namespace-workloads.json",
-    "nodeexporter.dashboard.url": "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/infrastructure/nodeexporter-nodes.json",
-    "nodes.dashboard.url": "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/infrastructure/nodes.json",
-    "workloads.dashboard.url": "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/infrastructure/workloads.json",
+    "grafanaGitOpsConfig": {
```
Comment (shapirov103) [inline cdk.json:20]: What is the reason for this change? Just general refactoring to reduce code duplication?

**Review by shapirov103 — COMMENTED** [review]
@freschri great work,  I added a few questions/comments. 
I understand this has dependency so can't be merged yet.

File: cdk.json (hunk 2 above)
Comment (freschri) [reply cdk.json:20]: basically yes: before this change, if I needed to e.g. like in this case change the flux path, I needed to duplicate the index file into another pattern and edit the single line..

File: cdk.json (hunk 1 above)
Comment (freschri) [reply cdk.json:29]: urls are copied from the existing ones, no change, I guess you are asking for an improvement?
I dont think it is possible anyway, because the urls are used in the GrafanaDashboard manifest for the grafana operator, e.g. see https://github.com/freschri/aws-observability-accelerator/blob/java-pattern/artifacts/grafana-operator-manifests/eks/java/amg_grafana-dashboards.yaml

File: lib/common/addons/amp-rules-configurator/amp-rules-configurator-addon.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    /** 
+     * Paths of the files listing the AMP rules.
+     */
+    ruleFilePaths: string[];
+}
+
+/**
+ * Implementation of Rules configuration for AMP
+ */
+export class AmpRulesConfiguratorAddOn implements blueprints.ClusterAddOn {
```
[the remark this answers is not in the document]
Comment (freschri) [reply lib/common/addons/amp-rules-configurator/amp-rules-configurator-addon.ts:25]: please see reply here: https://github.com/aws-observability/cdk-aws-observability-accelerator/pull/65#discussion_r1269145346

File: lib/common/addons/amp-rules-configurator/amp-rules-configurator-addon.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    readonly ampRulesConfiguratorAddOn: AmpRulesConfiguratorAddOnProps;
+    private _ruleGroupsNamespaces: CfnRuleGroupsNamespace[] = [];
```
[the remark this answers is not in the document]
Comment (freschri) [reply lib/common/addons/amp-rules-configurator/amp-rules-configurator-addon.ts]: done

File: cdk.json (hunk 2 above)
Comment (elamaran11) [reply cdk.json:20]: I like this approach.

File: docs/patterns/single-new-eks-observability-accelerators/single-new-eks-java-opensource-observability.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
+# Single New EKS Cluster Open Source Observability Accelerator - monitoring Java applications
```
Comment (elamaran11) [inline docs/patterns/single-new-eks-observability-accelerators/single-new-eks-java-opensource-observability.md:1]: Where is the pattern code for this doc?

**Review by elamaran11 — COMMENTED** [review]
@freschri I love the PR, great work for sure. I appreciate the fact that you have also added approach to refactor the context. I have following feedback :
1. I dont see a real pattern for Java workload monitoring. All i see is a doc showing the approach. We want a real pattern for our customers to try in single click and thats the purpose of these accelerators. I propose two approaches: first is creating a flag in context which says Java_Mon: Yes which will also apply java mon to `existing` and `single-new` open source patterns or second approach is to create all new pattern for existing or new single java monitoring for Open source so people will have clarity.
2. why do we see files for alerting-rules and recording-rules in this PR?

**Review by elamaran11 — CHANGES_REQUESTED** [review]
@freschri I love the PR, great work for sure. I appreciate the fact that you have also added approach to refactor the context. I have following feedback :

I dont see a real pattern for Java workload monitoring. All i see is a doc showing the approach. We want a real pattern for our customers to try in single click and thats the purpose of these accelerators. I propose two approaches: first is creating a flag in context which says Java_Mon: Yes which will also apply java mon to existing and single-new open source patterns or second approach is to create all new pattern for existing or new single java monitoring for Open source so people will have clarity.
why do we see files for alerting-rules and recording-rules in this PR?

Also fix the GH Warnings

**Comment by elamaran11** [comment]
@freschri Do we have an update on this PR?

**Comment by freschri** [comment]
@elamaran11 I added a flag to enable Java observability. Updated doc too. 2 dependencies on quickstart PRs:
https://github.com/aws-quickstart/cdk-eks-blueprints/pull/796
https://github.com/aws-quickstart/cdk-eks-blueprints/pull/776

**Comment by freschri** [comment]
> @freschri Do we have an update on this PR?

@elamaran11 I added a flag to enable Java observability. Updated doc too. 2 dependencies on quickstart PRs:
https://github.com/aws-quickstart/cdk-eks-blueprints/pull/796
https://github.com/aws-quickstart/cdk-eks-blueprints/pull/776

File: docs/patterns/single-new-eks-observability-accelerators/single-new-eks-java-opensource-observability.md (hunk 5 above)
Comment (freschri) [reply docs/patterns/single-new-eks-observability-accelerators/single-new-eks-java-opensource-observability.md:1]: small code changes to support new java flag and cdk file config. main code changes are in otel collector config (see quickstart PR).

**Comment by freschri** [comment]
> @freschri I love the PR, great work for sure. I appreciate the fact that you have also added approach to refactor the context. I have following feedback :
> 
>     1. I dont see a real pattern for Java workload monitoring. All i see is a doc showing the approach. We want a real pattern for our customers to try in single click and thats the purpose of these accelerators. I propose two approaches: first is creating a flag in context which says Java_Mon: Yes which will also apply java mon to `existing` and `single-new` open source patterns or second approach is to create all new pattern for existing or new single java monitoring for Open source so people will have clarity.
> 
>     2. why do we see files for alerting-rules and recording-rules in this PR?

Because there are dashboards in the pattern which need fixing. recording rules fix the the dashboards.

**Comment by elamaran11** [comment]
@freschri Please fix the GH Action issues.

**Review by elamaran11 — CHANGES_REQUESTED** [review]
@freschri This is AWSome work, thankyou for doing this. This a great new feature addition. I think approach is good, have some minor tactical feedback to fix the gaps :
1. Please merge from main, some of the changes like version, blueprints version is addressed.
2. AMP Alerting and recording rules are missing for Java Mon please AMP Alerting Rules and Recording rules are missing for Java Monitoring. Check [Terraform](https://github.com/aws-observability/terraform-aws-observability-accelerator/blob/main/modules/eks-monitoring/patterns/java/main.tf) for reference terraform reference.
3. A Separate documentation is needed in existing pattern too for Java monitoring.
4. Java mon code should be added to Graviton OSS pattern and exising OSS pattern

File: docs/patterns/single-new-eks-observability-accelerators/single-new-eks-java-opensource-observability.md [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        {
+          "kustomizationPath": "./artifacts/grafana-operator-manifests/eks/java"
+        }
+      ]
+    },
+    "java.pattern.enabled": true
+  }
+```
+
+Please also replace the `AmpAddOn` initialization in the `index.ts` file with:
```
[the remark this answers is not in the document]
Comment (freschri) [reply docs/patterns/single-new-eks-observability-accelerators/single-new-eks-java-opensource-observability.md]: yes thanks updated

File: lib/single-new-eks-awsnative-observability-construct/index.ts [hunk 7]
```diff
@@ -25,6 +25,7 @@ export default class SingleNewEksClusterAWSNativeobservabilityConstruct {
         ObservabilityBuilder.builder()
             .account(account)
             .region(region)
+            .version("auto")
```
[the remark this answers is not in the document]
Comment (freschri) [reply lib/single-new-eks-awsnative-observability-construct/index.ts]: done

File: lib/single-new-eks-cluster-construct/index.ts [hunk 8]
```diff
@@ -17,6 +17,7 @@ export default class SingleNewEksConstruct {
         ObservabilityBuilder.builder()
             .account(account)
             .region(region)
+            .version("auto")
```
[the remark this answers is not in the document]
Comment (freschri) [reply lib/single-new-eks-cluster-construct/index.ts]: done

File: lib/single-new-eks-mixed-observability-construct/index.ts [hunk 9]
```diff
@@ -34,6 +34,7 @@ export default class SingleNewEksMixedobservabilityConstruct {
         ObservabilityBuilder.builder()
             .account(account)
             .region(region)
+            .version("auto")
```
[the remark this answers is not in the document]
Comment (freschri) [reply lib/single-new-eks-mixed-observability-construct/index.ts]: done

File: lib/single-new-eks-opensource-observability-construct/index.ts [hunk 10]
```diff
@@ -14,20 +13,40 @@ export default class SingleNewEksOpenSourceobservabilityConstruct {
...
+                ruleFilePaths: [
+                    __dirname + '/../common/resources/amp-config/alerting-rules.yml',
+                    __dirname + '/../common/resources/amp-config/recording-rules.yml'
+                ]
+            }
+        };
 
+        if (utils.valueFromContext(scope, "java.pattern.enabled", false)) {
+            ampAddOnProps.openTelemetryCollector = {
+                manifestPath: __dirname + '/../common/resources/otel-collector-config.yml',
```
[the remark this answers is not in the document]
Comment (freschri) [reply lib/single-new-eks-opensource-observability-construct/index.ts]: yes thanks changed

File: lib/single-new-eks-opensource-observability-construct/graviton-index.ts [hunk 11]
```diff
@@ -38,33 +37,21 @@ export default class SingleNewEksGravitonOpenSourceObservabilityConstruct {
...
             new blueprints.addons.AmpAddOn({
-                ampPrometheusEndpoint: ampPrometheusEndpoint,
+                ampPrometheusEndpoint: ampEndpoint,
+                ampRules: {
+                    ampWorkspaceArn: ampWorkspaceArn,
+                    ruleFilePaths: [
+                        __dirname + '/../common/resources/amp-config/alerting-rules.yml',
+                        __dirname + '/../common/resources/amp-config/recording-rules.yml'
+                    ]
+                }
```
[the remark this answers is not in the document]
Comment (freschri) [reply lib/single-new-eks-opensource-observability-construct/graviton-index.ts]: done, doc updated too

File: lib/existing-eks-opensource-observability-construct/index.ts [hunk 12]
```diff
@@ -54,33 +53,21 @@ export default class ExistingEksOpenSourceobservabilityConstruct {
...
-                    "AMG_ENDPOINT_URL": amgEndpointUrl,
-                    "GRAFANA_CLUSTER_DASH_URL" : clusterDashUrl,
-                    "GRAFANA_KUBELET_DASH_URL" : kubeletDashUrl,
-                    "GRAFANA_NSWRKLDS_DASH_URL" : namespaceWorkloadsDashUrl,
-                    "GRAFANA_NODEEXP_DASH_URL" : nodeExporterDashUrl,
-                    "GRAFANA_NODES_DASH_URL" : nodesDashUrl,
-                    "GRAFANA_WORKLOADS_DASH_URL" : workloadsDashUrl
-                },
-            }),
+            new blueprints.addons.FluxCDAddOn({"repositories": [fluxRepository]}),
```
[the remark this answers is not in the document]
Comment (freschri) [reply lib/existing-eks-opensource-observability-construct/index.ts]: done

File: lib/existing-eks-opensource-observability-construct/index.ts [hunk 13]
```diff
@@ -54,33 +53,21 @@ export default class ExistingEksOpenSourceobservabilityConstruct {
...
-                ampPrometheusEndpoint: ampPrometheusEndpoint,
+                ampPrometheusEndpoint: ampEndpoint,
+                ampRules: {
+                    ampWorkspaceArn: ampWorkspaceArn,
+                    ruleFilePaths: [
+                        __dirname + '/../common/resources/amp-config/alerting-rules.yml',
+                        __dirname + '/../common/resources/amp-config/recording-rules.yml'
+                    ]
+                }
             }),
```
[the remark this answers is not in the document]
Comment (freschri) [reply lib/existing-eks-opensource-observability-construct/index.ts]: done

File: lib/existing-eks-opensource-observability-pattern/index.ts [hunk 14]
```diff
@@ -39,12 +40,35 @@ export default class ExistingEksOpenSourceobservabilityPattern {
...
+            ampRules: {
+                ampWorkspaceArn: ampWorkspaceArn,
+                ruleFilePaths: [
+                    __dirname + '/../common/resources/amp-config/alerting-rules.yml',
+                    __dirname + '/../common/resources/amp-config/recording-rules.yml'
+                ]
+            }
+        };
+
+        if (utils.valueFromContext(scope, "java.pattern.enabled", false)) {
```
Comment (elamaran11) [inline lib/existing-eks-opensource-observability-pattern/index.ts:59]: `(utils.valueFromContext(scope, "java.pattern.enabled", undefined)` situation is not handled. I think people may ignore this in few patterns.

File: lib/single-new-eks-opensource-observability-pattern/graviton-index.ts [hunk 15]
```diff
@@ -15,17 +15,41 @@ export default class SingleNewEksGravitonOpenSourceObservabilityPattern {
...
+            ampRules: {
+                ampWorkspaceArn: ampWorkspaceArn,
+                ruleFilePaths: [
+                    __dirname + '/../common/resources/amp-config/alerting-rules.yml',
+                    __dirname + '/../common/resources/amp-config/recording-rules.yml'
+                ]
+            }
+        };
+
+        if (utils.valueFromContext(scope, "java.pattern.enabled", false)) {
```
Comment (elamaran11) [inline lib/single-new-eks-opensource-observability-pattern/graviton-index.ts:40]: (utils.valueFromContext(scope, "java.pattern.enabled", undefined) situation is not handled. I think people may ignore this in few patterns.

File: lib/single-new-eks-opensource-observability-pattern/index.ts [hunk 16]
```diff
@@ -14,20 +14,44 @@ export default class SingleNewEksOpenSourceobservabilityPattern {
...
+            ampRules: {
+                ampWorkspaceArn: ampWorkspaceArn,
+                ruleFilePaths: [
+                    __dirname + '/../common/resources/amp-config/alerting-rules.yml',
+                    __dirname + '/../common/resources/amp-config/recording-rules.yml'
+                ]
+            }
+        };
+
+        if (utils.valueFromContext(scope, "java.pattern.enabled", false)) {
```
Comment (elamaran11) [inline lib/single-new-eks-opensource-observability-pattern/index.ts:42]: (utils.valueFromContext(scope, "java.pattern.enabled", undefined) situation is not handled. I think people may ignore this in few patterns.

**Review by elamaran11 — COMMENTED** [review]
@freschri Everything looks great, just minor error handling is missed. This is fixing so many issues btw. Im currently testing it locally, will let you know the outcome. Also will run the e2e here.

**Comment by elamaran11** [comment]
@freschri I was able to run this locally and i can see all Dashboards showing data, great work. Once you fix the above issues and resolve both the commands, we can proceed to merge.

File: lib/existing-eks-opensource-observability-pattern/index.ts (hunk 14 above)
Comment (freschri) [reply lib/existing-eks-opensource-observability-pattern/index.ts:59]: false is the default value is not found. I guess no change here right?

File: lib/single-new-eks-opensource-observability-pattern/graviton-index.ts (hunk 15 above)
Comment (freschri) [reply lib/single-new-eks-opensource-observability-pattern/graviton-index.ts:40]: same as above

File: lib/single-new-eks-opensource-observability-pattern/index.ts (hunk 16 above)
Comment (freschri) [reply lib/single-new-eks-opensource-observability-pattern/index.ts:42]: same as above

File: lib/existing-eks-opensource-observability-pattern/index.ts (hunk 14 above)
Comment (elamaran11) [reply lib/existing-eks-opensource-observability-pattern/index.ts:59]: I think i agree.

**Comment by elamaran11** [comment]
@freschri Please merge from Main, some work from upstream is missing. Once done i can merge.

**Review by elamaran11 — APPROVED** [review]
LGTM. Nice work @freschri
